README.md
# Chatango HTML Mini Overlay #

[Chatango](http://chatango.com) is a basic chat site, with a user interface
based around HTML5 and JavaScript. However, before the HTML5/JS interface
was created, the interface was made with Flash.  
This version had one major feature. Many users still use the outdated Flash
version because this feature is missing from the new version.  
The feature is a simple overlay, which appears near the cursor when
you hover over another user's profile picture. This overlay would show
information about the user. Their age, gender, location, and a small description
would appear here if they added it to their profile. As well as an enlarged
version of their thumbnail.

I was bored one day, and decided to figure out how to make a UserScript.
I decided to try to make my own overlay for the HTML5 chat, to fill in for
the one that was not added by the developers (yet, probably).

## IMPORTANT NOTES ##
I __do not__ claim this script to be perfect, never have bugs, and work for
every user's mini profile. It does not, and the amount of work required to
make sure every mini fits is not even worth it, if even possible, and would
create more bugs that would have to be fixed. And in fixing those bugs, other
stuff will break. I will make attempts to improve it over time, but I will not
try for perfection, unless I find a new method which works perfectly.

I cannot anticipate what users may put on their profile, and I am not sure
to what extent Chatango filters the profile text. The script will try it's best
to remove any (potentially) harmful code from the profile before showing it.
If you do find a security bug with this, __PLEASE__ let me know immediately
so I can get out a fix as soon as possible.

Not all profiles will work with this. However, it supports __far more__ than
the official flash version does, and shows each profile in more detail.

This is not intended to be an entire profile viewer in the first place. It is
meant to be more of a preview. A much larger preview than the flash version
provides.

This is an early release. I tried to squash any bugs I could find, but I am
sure there will be some that I missed, or that I never noticed, or that only
happens on a browser I did not test on.  
If you find a bug, or have a suggestion. Please post a new issue.  

If you know your way around JavaScript, HTML, and CSS. Feel free to edit your
local copy of the script to match the colors, design, etc. to match your
preferences.  
I do plan to make many of the aspects configurable, with time.

Because of the strange way Chatango stores the user's age, it's highly possible
that the number shown may be off by a year. This will be fixed in the future.

## Benefits over the flash version ##
- The overlay will automatically scale to fit the window and content. If your
window is wide and tall, the overlay will be wider, and the maximum height will
be taller.
- More of the user's mini profile will be shown. The content shrinks to fit
smaller windows. It however, does not grow to fill the entire possible area.
- The profile image shown is the user's full image, not the thumbnail, and
scales to the original dimensions. Unless you have high-speed/fiber/etc.
internet, this __will__ cause a slower loading time on the images.

## Reporting bugs ##
If you do find a bug, please let me know about it by posting a new issue.  
When posting an issue, include the name and version (if possible) of your
browser, as well as the name of the UserScript manager (GreaseMonkey,
TamperMonkey, etc.) that you are using.

## Suggestions? Questions? Feedback? ##
Many users have no idea how important these can be to a developer!

You can post an issue for suggestions, questions, and feedback alike. They
don't need to be actual issues!  
It's recommended to tag these, but if you don't know how, that's perfectly
fine. Better to post it with no tags than to not post it at all.

If you'd rather not post a public issue, I'd be glad to hear your
feedback/questions/suggestions. Send me a Private Message on Chatango, my
main account is [Hazerd](http://hazerd.chatango.com). I always keep this
account on the chatango app, so I get instant notifications. If I don't reply
right away, I'll get to it as soon as I have a chance. If it's been a while,
just try re-sending the message. Odds are I just never got it.

# Installation #

1. Install a UserScript manager for your browser.
[GreaseMonkey](http://www.greasespot.net/) and
[TamperMonkey](http://tampermonkey.net/) are two popular choices.
2. Visit the [Greasy Fork Page](https://greasyfork.org/en/scripts/13044-chatango-html-mini-overlay)
and click the Green "Install this script" button.
3. This step varies on your manager. But generally, there should be a
button to Install the script.


CHANGELOG.md
# Changelog #

### Sometime In The Future ###
- Make sure the A;S;L box and the content box are always the same width.
- Allow configuration of several features, colors, sizes, etc.

### v0.3 ###
- Removed the scrollbar.

### v0.2 ###
- Added CHANGELOG.
- Updated README.
- Scripts are automatically deleted from the mini profile text before adding
it to the box, for security reasons.
- Changed the commas (`,`) in the ASL bar to semicolons (`;`), because it
looks confusing when the location is present as it includes another comma.
- Content box and ASL now auto-scale to the content, with a set minimum and
maximum width. If the maximum width is reached and the content is still too
large, then the content gets scaled to fit (as much as possible due to browser
limitations).
- Cleaned up the code a bit.

### v0.1 ###
- Initial Release
- Posted to GitHub and GreasyFork


Chatango-HTML-Mini-Popup.user.js
// ==UserScript==
// @name         Chatango HTML Mini Overlay
// @namespace    Chatango-HTML-Mini-Overlay
// @version      0.1
// @description  An extension for chatango.com's new html5/js chat. Which adds the old "mini profile on hover" feature.
// @author       Hazerd (Hazerdous) <Daniel Ransom>
// @homepage     http://hazerd.com
// @support      https://github.com/Hazerdous/Chatango-HTML-Mini-Overlay/issues
// @include      http://*.chatango.com/*
// @include      https://*.chatango.com/*
// @require      http://code.jquery.com/jquery-1.11.3.js
// @grant        GM_xmlhttpRequest
// ==/UserScript==

'use strict';
$("body").append(
'<div id="floating-mini-container"\
style="position:absolute; width:60%; min-width:300px; border-spacing:0; overflow:hidden;\
border-style:solid; border-width:1px; z-index:999; background-color:rgba(100,100,100,0.9); visibility:hidden;" >\
<div id="floating-mini-content"\
style="height:100%;width:100%; white-space:pre; overflow:scroll;\
text-shadow:1px 0 black, -1px 0 black, 0 1px black, 0 -1px black; color:white;">\
</div>\
</div>\
<img id="floating-mini-image" src="" alt="Loading . . ."\
style="position:absolute; max-height:150px; border-style:solid; border-width:1px;\
z-index:1000; visibility:hidden;">\
<div id="floating-mini-asl"\
style="position:absolute; white-space:nowrap; min-width:300px; width:60%; height:20px; font-size:20px; line-height:1;\
border-style:solid; border-width:1px; border-color:black; z-index:1000; background-color:rgba(100,100,100,0.9); visibility:hidden;\
text-shadow:1px 0 black, -1px 0 black, 0 1px black, 0 -1px black; color:white;">'
)

var elem = $("#floating-mini-container")
var image = $("#floating-mini-image")
var content = $("#floating-mini-content")
var asl = $("#floating-mini-asl")
var visible = false
var ASL = ["Gender?", "Age?", "Location?"]
var cursorX = 0
var cursorY = 0

elem.css("max-height",(window.innerHeight - 174).toString() + "px")

$(document).on("mouseenter", ".user-thumb", function(e) {
 var name = this.classList[1].match(/thumb-image-(.+)$/)[1]
 image[0].src="http://fp.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/full.jpg"
 getUserData(name)
})
$(document).on("mousemove", function(e) {
 cursorX = e.pageX
 cursorY = e.pageY
 if (e.target.classList[0] == "user-thumb") {
  if (elem.css("visibility") == "hidden" || image.css("visibility") == "hidden" || asl.css("visibility") == "hidden") {
   visible = true
   asl.css("visibility", "visible")
   elem.css("visibility", "visible")
   image.css("visibility", "visible")
  }
 } else {
  if (elem.css("visibility") == "visible" || image.css("visibility") == "visible" || asl.css("visibility") == "visible") {
   visible = false
   asl.css("visibility", "hidden")
   elem.css("visibility", "hidden")
   image.css("visibility", "hidden")
   content[0].innerHTML = "Loading . . ."
   resetASL()
   image[0].src= ""
   content[0].style.zoom = 1
  }
 }
 updatePosition()

})

$(window).resize(function() {
 elem.css("max-height", (window.innerHeight - 300).toString() + "px")
})

function updatePosition() {
 var offsetY = 0
 if (elem[0].offsetHeight + cursorY + 100 > window.innerHeight) {
  offsetY = window.innerHeight - (elem[0].offsetHeight + cursorY + 100)
 }
 elem.css({left:cursorX+25,top:cursorY+100+offsetY})
 image.css({left:cursorX+25,bottom:window.innerHeight-cursorY-80-offsetY})
 asl.css({left:cursorX+25,bottom:window.innerHeight-cursorY-101-offsetY})
}
function getUserData(name) {
 GM_xmlhttpRequest({
  method: "GET",
  url: "http://st.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/mod1.xml",
  onload: function(res) {
   extractASL(res.response)
   var body = res.response.match(/<body>(.*)<\/body>/)
   if (body !== null) { body = body[1] }
   setAbout(body)
  }
 })
}

function setImage(name) {
 image[0].src="http://fp.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/full.jpg"
}
function setAbout(text) {
 if (text === null) {
  content[0].innerHTML = "No Profile Data"
 } else {
  content[0].innerHTML = decodeURIComponent(text.replace(/%0D/g, "<br>"))
  content[0].style.zoom = elem[0].clientWidth / content[0].scrollWidth
  elem[0].style.height = content[0].scrollheight * content[0].style.zoom
 }
 updatePosition()
}
function setAge(age) {
 if (age === null) { ASL[0] = "Age?" }
 else { ASL[0] = age.toString() }
}
function setGender(gen) {
 if (gen == "M") {
  ASL[1] = "Male"
 } else if (gen == "F") {
  ASL[1] = "Female"
 } else {
  ASL[1] = "Gender?"
 }
}
function setLocation(loc) {
 ASL[2] = loc || "Location?"
}
function setASL(a, s, l) {
 setAge(a)
 setGender(s)
 setLocation(l)
 asl[0].innerHTML = ASL.join(", ")
}
function extractASL(data) {
 var a = data.match(/<b>(\d\d\d\d)-\d\d-\d\d<\/b>/)
 if (a) { a = new Date().getFullYear() - parseInt(a[1]) }
 var s = data.match(/<s>(\w)<\/s>/)
 if (s) { s = s[1] }
 var l = data.match(/<l .*?>(.*?)<\/l>/)
 if (l) { l = l[1] }
 setASL(a, s, l)
}
function resetASL() {
 ASL = ["Gender?", "Age?", "Location?"]
 asl[0].innerHTML = "Loading . . ."
}


Chatango-HTML-Mini-Overlay.user.js
// ==UserScript==
// @name         Chatango HTML Mini Overlay
// @namespace    Chatango-HTML-Mini-Overlay
// @version      0.2
// @description  An extension for chatango.com's new html5/js chat. Which adds the old "mini profile on hover" feature.
// @author       Hazerd (Hazerdous) <Daniel Ransom>
// @homepage     https://github.com/Hazerdous/Chatango-HTML-Mini-Overlay/
// @support      https://github.com/Hazerdous/Chatango-HTML-Mini-Overlay/issues
// @include      http://*.chatango.com/*
// @include      https://*.chatango.com/*
// @require      http://code.jquery.com/jquery-1.11.3.js
// @grant        GM_xmlhttpRequest
// ==/UserScript==

'use strict';
$("body").append(
'\
<div id="floating-mini-container"\
style="position:absolute; max-width:80%; min-width:300px; border-spacing:0;\
overflow:hidden; border-style:solid; border-width:1px; z-index:999;\
background-color:rgba(100,100,100,0.9); visibility:hidden;" >\
<div id="floating-mini-content"\
style="height:100%;width:100%; white-space:pre; overflow:hidden; color:white;\
text-shadow:1px 0 black, -1px 0 black, 0 1px black, 0 -1px black;">\
</div>\
</div>\
<img id="floating-mini-image" src="" alt="Loading . . ."\
style="position:absolute; max-height:150px; border-style:solid; z-index:1000;\
border-width:1px; visibility:hidden;">\
<div id="floating-mini-asl"\
style="position:absolute; white-space:nowrap; min-width:300px; max-width:80%;\
height:20px; font-size:20px; line-height:1; border-style:solid; z-index:1000;\
border-width:1px; border-color:black; visibility:hidden; color:white;\
background-color:rgba(100,100,100,0.9);\
text-shadow:1px 0 black, -1px 0 black, 0 1px black, 0 -1px black;">\
'
)

var elem = $("#floating-mini-container")
var image = $("#floating-mini-image")
var content = $("#floating-mini-content")
var asl = $("#floating-mini-asl")
var visible = false
var ASL = ["Gender?", "Age?", "Location?"]
var cursorX = 0
var cursorY = 0

elem.css("max-height",(window.innerHeight - 174).toString() + "px")

$(document).on("mouseenter", ".user-thumb", function(e) {
 var name = this.classList[1].match(/thumb-image-(.+)$/)[1]
 image[0].src="http://fp.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/full.jpg"
 getUserData(name)
})
$(document).on("mousemove", function(e) {
 cursorX = e.pageX
 cursorY = e.pageY
 if (e.target.classList[0] == "user-thumb") {
  if (elem.css("visibility") == "hidden" || image.css("visibility") == "hidden" || asl.css("visibility") == "hidden") {
   visible = true
   asl.css("visibility", "visible")
   elem.css("visibility", "visible")
   image.css("visibility", "visible")
  }
 } else {
  if (elem.css("visibility") == "visible" || image.css("visibility") == "visible" || asl.css("visibility") == "visible") {
   visible = false
   asl.css("visibility", "hidden")
   elem.css("visibility", "hidden")
   image.css("visibility", "hidden")
   setAbout()
   resetASL()
   setImage()
   content[0].style.zoom = 1
  }
 }
 updatePosition()

})

$(window).resize(function() {
 elem.css("max-height", (window.innerHeight - 300).toString() + "px")
})

function updatePosition() {
 var offsetY = 0
 if (elem[0].offsetHeight + cursorY + 100 > window.innerHeight) {
  offsetY = window.innerHeight - (elem[0].offsetHeight + cursorY + 100)
 }
 elem.css({left:cursorX+25,top:cursorY+100+offsetY})
 image.css({left:cursorX+25,bottom:window.innerHeight-cursorY-80-offsetY})
 asl.css({left:cursorX+25,bottom:window.innerHeight-cursorY-101-offsetY})
}
function getUserData(name) {
 GM_xmlhttpRequest({
  method: "GET",
  url: "http://st.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/mod1.xml",
  onload: function(res) {
   extractASL(res.response)
   var body = res.response.match(/<body>(.*)<\/body>/)
   if (body !== null) { body = body[1] }
   setAbout(body)
  }
 })
}

function setImage(name) {
 if (!name) {
  image[0].src=""
 } else {
  image[0].src="http://fp.chatango.com/profileimg/" + name[0] + "/" + (name[1] || name[0]) + "/" + name + "/full.jpg"
 }
}
function setAbout(text) {
 if (text === undefined) {
  content[0].innerHTML = "Loading . . ."
 } else if (text === null) {
  content[0].innerHTML = "No Profile Data"
 } else {
  content[0].innerHTML = decodeURIComponent(text.replace(/%0D/g, "<br>")).replace(/<[^>]*?script.*?>.*?<[^>]*?\/script.*?>/gi, "")
  content[0].style.zoom = elem[0].clientWidth / content[0].scrollWidth
  elem[0].style.height = content[0].scrollheight * content[0].style.zoom
 }
 updatePosition()
}
function setAge(age) {
 if (!age) {
  ASL[0] = "Age?"
 } else {
  ASL[0] = age.toString()
 }
}
function setGender(gen) {
 if (gen == "M") {
  ASL[1] = "Male"
 } else if (gen == "F") {
  ASL[1] = "Female"
 } else {
  ASL[1] = "Gender?"
 }
}
function setLocation(loc) {
 if (!loc) {
  ASL[2] = "Location?"
 } else {
  ASL[2] = loc
 }
}
function setASL(a, s, l) {
 setAge(a)
 setGender(s)
 setLocation(l)
 asl[0].innerHTML = ASL.join("; ")
}
function extractASL(data) {
 var a = data.match(/<b>(\d\d\d\d)-\d\d-\d\d<\/b>/)
 if (a) { a = new Date().getFullYear() - parseInt(a[1]) }
 var s = data.match(/<s>(\w)<\/s>/)
 if (s) { s = s[1] }
 var l = data.match(/<l .*?>(.*?)<\/l>/)
 if (l) { l = l[1] }
 setASL(a, s, l)
}
function resetASL() {
 ASL = ["Gender?", "Age?", "Location?"]
 asl[0].innerHTML = "Loading . . ."
}
